packages/accounts-password/password_server.js

```diff
@@ -573,119 +573,201 @@ Accounts.sendResetPasswordEmail = function (userId, email) {
+ * @summary Generates an e-mail verification token record.
  * @locus Server
- * @param {String} userId The id of the user to send email to.
- * @param {String} [email] Optional. Which address of the user's to send the email to. This address must be in the user's `emails` list. Defaults to the first email in the list.
+ * @param {String} userId The id of the user to generate the  e-mail verification token record for.
+ * @param {String} email Which address of the user to generate the e-mail verification token record for. This address must be in the user's `emails` list. If `null`, defaults to the first unverified email in the list.
+ * @param {Object} [extraTokenData] Optional additional data to be added into the token record.
+ * @returns {Object} Object with {email, user, tokenRecord} values.
  * @importFromPackage accounts-base
  */
-Accounts.sendEnrollmentEmail = function (userId, email) {
-  // XXX refactor! This is basically identical to sendResetPasswordEmail.
-
-  // Make sure the user exists, and email is in their addresses.
+Accounts.generateVerificationTokenRecord = function (userId, email, extraTokenData) {
+  // Make sure the user exists, and email is one of their addresses.
   var user = Meteor.users.findOne(userId);
   if (!user) {
-    throw new Error("Can't find user");
+    handleError("Can't find user");
   }
-  // pick the first email if we weren't passed an email.
-  if (!email && user.emails && user.emails[0]) {
-    email = user.emails[0].address;
+
+  // pick the first unverified email if we weren't passed an email.
+  if (!email) {
+    var emailRecord = _.find(user.emails || [], function (e) { return !e.verified; });
+    email = (emailRecord || {}).address;
+
+    if (!email) {
+      handleError("That user has no unverified email addresses.");
+    }
   }
+
   // make sure we have a valid email
   if (!email || !_.contains(_.pluck(user.emails || [], 'address'), email)) {
-    throw new Error("No such email for user.");
+    handleError("No such email for user.");
   }
 
-  var token = Random.secret();
-  var when = new Date();
   var tokenRecord = {
-    token: token,
-    email: email,
-    when: when,
-    reason: 'enroll'
+    token: Random.secret(),
+    // TODO: This should probably be renamed to "email" to match reset token record.
+    address: email,
+    when: new Date()
   };
-  Meteor.users.update(userId, {$set: {
-    "services.password.reset": tokenRecord
+
+  if (extraTokenData) {
+    _.extend(tokenRecord, extraTokenData);
+  }
+
+  return {email, user, tokenRecord};
+};
+
+/**
+ * @summary Saves the reset token record into database in user's document.
+ * @param {Object} user User for which to save the reset token record for.
+ * @param {Object} tokenRecord The reset token record to be saved.
+ * @returns {Object} User object with added token record.
+ * @importFromPackage accounts-base
+ */
+Accounts.saveResetTokenRecord = function (user, tokenRecord) {
```

Comment:
This seems like it should be private and handled within the `generate*` methods, as we originally discussed.  I don't recall a reason to make multiple new public APIs.